src/test/scala/org/ergoplatform/tools/ChainGenerator.scala

```diff
@@ -0,0 +1,96 @@
+package org.ergoplatform.tools
+
+import java.io.File
+
+import org.ergoplatform.mining.{CandidateBlock, EquihashPowScheme}
+import org.ergoplatform.modifiers.ErgoFullBlock
+import org.ergoplatform.modifiers.history.{Extension, ExtensionCandidate, Header}
+import org.ergoplatform.nodeView.history.ErgoHistory
+import org.ergoplatform.nodeView.state.StateType
+import org.ergoplatform.settings._
+import org.ergoplatform.utils.{ErgoTestHelpers, ValidBlocksGenerators}
+import scorex.core.utils.ScorexLogging
+import scorex.utils.Random
+
+import scala.annotation.tailrec
+import scala.concurrent.duration._
+
+/**
+  * Application object for chain generation.
+  * Takes 2 parameters: start timestamp and path to history folder.
+  * Generate blocks starting from start timestamp and until current time with expected block interval
+  * between them, to ensure that difficulty does not change.
+  */
+object ChainGenerator extends App with ValidBlocksGenerators with ErgoTestHelpers with ScorexLogging {
+
+  val n: Char = 96
+  val k: Char = 5
+  val pow = new EquihashPowScheme(n, k)
+  val blockInterval = 2.minute
+
+  val startTime = args.headOption.map(_.toLong).getOrElse(timeProvider.time - (blockInterval * 10).toMillis)
+  val dir = if (args.length < 2) new File("/tmp/ergo/node1/data") else new File(args(1))
+  dir.mkdirs()
+
+  val history = {
+    val miningDelay = 1.second
+    val minimalSuffix = 2
+    val nodeSettings: NodeConfigurationSettings = NodeConfigurationSettings(StateType.Utxo, verifyTransactions = true,
+      -1, PoPoWBootstrap = false, minimalSuffix, mining = false, miningDelay, offlineGeneration = false, 200)
+    val chainSettings = ChainSettings(blockInterval, 256, 8, pow,
+      settings.chainSettings.monetary)
+    val fullHistorySettings: ErgoSettings = ErgoSettings(dir.getAbsolutePath, chainSettings, settings.testingSettings,
+      nodeSettings, CacheSettings.default, settings.scorexSettings)
+    ErgoHistory.readOrGenerate(fullHistorySettings, timeProvider)
+  }
+  if (history.bestHeaderOpt.nonEmpty) {
+    log.error(s"History at ${dir.getAbsoluteFile} already exists. Exiting.")
+    System.exit(11)
+  }
+
+  var (state, boxHolder) = createUtxoState()
+
+  def loop(last: Option[Header], acc: Seq[ErgoFullBlock]): Seq[ErgoFullBlock] = {
+    val time: Long = last.map(_.timestamp + blockInterval.toMillis).getOrElse(startTime)
+    if (time < timeProvider.time) {
+      val txsBoxes = validTransactionsFromBoxHolder(boxHolder)
+      val txs = txsBoxes._1
+      boxHolder = txsBoxes._2
+
+      @tailrec
+      def generate(candidate: CandidateBlock): ErgoFullBlock = {
```

Comment:
this could be a top-level function as it looks not to use outer function context